invoking docc from toolchains

shilpeegupta14 opened the pull request:
Adjusted the generate_docc.sh file to invoke docc from toolchains now.

### Motivation:

Now that the DocC ships in toolchains, I adjusted the generate_docc.sh file to invoke docC directly from toolchains. I am making this change in order to learn contributing effectively. I am trying to update the DocC.

### Modifications:
1) changed the prepare and build toolchain(earlier line no 50 to 73) code to directly access docc from toolchains.
2) modified and wrote
     export TOOLCHAIN=/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-11-20-a.xctoolchain.  
     cd $TOOLCHAIN/usr/bin/docc

### Result:

- Resolves #856 
<img width="974" alt="Screenshot 2021-12-01 at 6 54 03 PM" src="https://user-images.githubusercontent.com/78029920/144242181-8536f3b6-303a-4a38-a0f6-f8384c025459.png">

ktoso commented on scripts/docs/generate_docc.sh:
```diff
@@ -47,30 +47,9 @@ declare -r build_path_linux="$build_path/"
-
-export DOCC_HTML_DIR=$docc_render_source_path/dist
+# Accessing docc from toolchain
+export TOOLCHAIN=/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-11-20-a.xctoolchain
```
I mean, no, we cannot hardcode the script to a specific path -- it should ask for the toolchain to be set instead for example -- require the TOOLCHAIN var exists.

shilpeegupta14 replied:
Something like this....

if  [[ ! -d "$docc_source_path" ]]; then 
  cd /Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-11-20-a.xctoolchain
else
  echo "TOOLCHAIN var exists."

heckj replied:
@shilpeegupta14 - I suspect what you want to do is check to see if the environment variable TOOLCHAIN is already set, and if so - use it. For a linux platform with a nightly swift pre-installed and already added to the relevant paths (such as a docker build from a nightly swift toolchain release), then you won't need to prefix the commands with TOOLCHAIN, as they *should* be already on the path (I haven't double-checked that hypothesis though).

In either case, you'll want to likely set the `DOCC_HTML_DIR` environment variable so that `docc`, however you invoke it, knows where to find its HTML goodies it wants to combine/link with the output.

[StackOverflow has a nice reference](https://stackoverflow.com/questions/3601515/how-to-check-if-a-variable-is-set-in-bash) to this otherwise awkward problem of determining if an environment variable is already set, which amounts to doing something like:
`if [ -z ${var+x} ]; then echo "var is unset"; else echo "var is set to '$var'"; fi`

Translating that to "require an environment TOOLCHAIN to be set in order to proceed" might be:

`if [ -z ${TOOLCHAIN+x} ]; then echo "TOOLCHAIN is unset"; exit(1); fi`

Then from there, build up the `DOCC_HTML_DIR`:

`export DOCC_HTML_DIR=${TOOLCHAIN}/usr/share/docc/render/`

ktoso commented on scripts/docs/generate_docc.sh:
```diff
@@ -47,30 +47,9 @@ declare -r build_path_linux="$build_path/"
-export DOCC_HTML_DIR=$docc_render_source_path/dist
+# Accessing docc from toolchain
+export TOOLCHAIN=/Library/Developer/Toolchains/swift-DEVELOPMENT-SNAPSHOT-2021-11-20-a.xctoolchain
+cd $TOOLCHAIN/usr/bin/docc
```
this is a binary, we don't "cd" it.

shilpeegupta14 replied:
okay so only this much will help
$TOOLCHAIN/usr/bin/docc

heckj replied:
yep, invoke with `$TOOLCHAIN/usr/bin/docc` once you've verified that TOOLCHAIN is defined.

shilpeegupta14 commented:
Adjusted the generate_docc.sh file to invoke docc from toolchains now.

**Motivation:**

Now that the DocC ships in toolchains, I adjusted the generate_docc.sh file to invoke docC directly from toolchains. I am making this change in order to learn contributing effectively. I am trying to update the DocC.

**Modifications:**

1) changed the prepare and build toolchain(earlier line no 50 to 73) code to directly access docc from toolchains.
2) modified and wrote
```
 if [ -z ${TOOLCHAIN+x} ]; then 
    echo "TOOLCHAIN is unset"; exit(1); 
 fi
 export DOCC_HTML_DIR = ${TOOLCHAIN}/usr/bin/docc

```
3) `$TOOLCHAIN/usr/bin/docc`
<img width="878" alt="Screenshot 2021-12-04 at 9 51 16 AM" src="https://user-images.githubusercontent.com/78029920/144696761-2eb196ba-9933-4509-b80f-cc7f831011fe.png">

shilpeegupta14 commented:
@swift-server-bot please test

Kyle-Ye commented on scripts/docs/generate_docc.sh:
```diff
@@ -47,30 +47,13 @@ declare -r build_path_linux="$build_path/"
-  echo "Assuming docc-render is built..."
+# invoking docc from toolchain
+if [-z ${TOOLCHAIN+x} ]; then   
+  echo "TOOLCHAIN is unset"; exit(1);
```
This gives an error on my machine with zsh.

line 52: syntax error near unexpected token `1'

Kyle-Ye commented on scripts/docs/generate_docc.sh:
```diff
@@ -47,30 +47,13 @@ declare -r build_path_linux="$build_path/"
+export DOCC_HTML_DIR=${TOOLCHAIN}/usr/bin/docc
 
-export DOCC_HTML_DIR=$docc_render_source_path/dist
+$TOOLCHAIN/usr/bin/docc
```
Calling docc seems useless here for me.

What we want is to export the env DOCC_HTML_DIR so that we can use it later when call docc in preview_docc.sh

Kyle-Ye commented on scripts/docs/generate_docc.sh:
```diff
@@ -47,30 +47,13 @@ declare -r build_path_linux="$build_path/"
+if [-z ${TOOLCHAIN+x} ]; then   
+  echo "TOOLCHAIN is unset"; exit(1);
 fi
+export DOCC_HTML_DIR=${TOOLCHAIN}/usr/bin/docc
```
DOCC_HTML_DIR should point to the docc-render dist dir or in the toolchain it is located at /usr/share/docc/render

ktoso commented:
Replaced by https://github.com/apple/swift-distributed-actors/pull/865

Thanks for the effort @shilpeegupta14, maybe you'll find something else you can contribute; 
Thanks @Kyle-Ye for the PR!
